{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Script 1:\n",
    "Including vcf data into genbank files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO,Seq\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import SeqFeature\n",
    "import vcf\n",
    "\n",
    "def vcftogenbank(reference, sample):\n",
    "    \n",
    "    genbank=SeqIO.read(reference,\"genbank\") #open reference genbank file\n",
    "    new_features=[]\n",
    "\n",
    "    vcf_file=vcf.Reader(filename=\"vcf_files/\"+sample+\"_c.vcf.gz\",compressed=\"True\") #open vcf file\n",
    "    new_sequence=\"\"\n",
    "    new_sequence=str(genbank.seq.strip())\n",
    "    old=len(new_sequence)\n",
    "\n",
    "\n",
    "    dif=0\n",
    "\n",
    "    for SNP in vcf_file:\n",
    "        new_sequence=new_sequence[:SNP.POS-1+dif]+str(SNP.ALT[0]).lower()+new_sequence[SNP.POS+len(SNP.REF)-1+dif:] #including sequence changes into new sequence\n",
    "        difdif=0\n",
    "        difdif=len(SNP.ALT[0])-len(SNP.REF) #in case of INDELs sequence positions need to be adapted\n",
    "        dif=dif+difdif\n",
    "\n",
    "        for feature in genbank.features: #updating features\n",
    "            if (feature.location.start.position > SNP.POS) and (feature.location.end.position > SNP.POS): #if SNP upstream of feature      \n",
    "                feature.location = SeqFeature.FeatureLocation(SeqFeature.ExactPosition(feature.location.start+difdif),\n",
    "                                                                     SeqFeature.ExactPosition(\n",
    "                                                                    feature.location.end+difdif),\n",
    "                                                                        feature.location.strand)\n",
    "            elif (feature.location.start.position <= SNP.POS) and (feature.location.end.position > SNP.POS): #if SNP is within feature\n",
    "                feature.location = SeqFeature.FeatureLocation(SeqFeature.ExactPosition(feature.location.start),\n",
    "                                                                     SeqFeature.ExactPosition(\n",
    "                                                                    feature.location.end+difdif),\n",
    "                                                                    feature.location.strand)\n",
    "                if \"translation\" in feature.qualifiers:\n",
    "                    transcript=\"\"\n",
    "                    translation=\"\"\n",
    "                    transcript=transcript+new_sequence[feature.location.start-1:feature.location.end] \n",
    "                    transcript=Seq(transcript) #getting feature transcript\n",
    "\n",
    "                    if feature.location.strand == -1:\n",
    "                        transcript=transcript.reverse_complement()\n",
    "\n",
    "                    translation=transcript.translate() #translate trancript and including in feature description\n",
    "                    feature.qualifiers[\"translation\"]=str(translation)\n",
    "\n",
    "\n",
    "            elif (feature.location.start.position < SNP.POS) and (feature.location.end.position >= SNP.POS): #if SNP is downstream of feature\n",
    "                feature.location = SeqFeature.FeatureLocation(SeqFeature.ExactPosition(feature.location.start),\n",
    "                                                                     SeqFeature.ExactPosition(\n",
    "                                                                    feature.location.end),\n",
    "                                                                        feature.location.strand)\n",
    "\n",
    "\n",
    "\n",
    "    for feature in genbank.features: #getting new features\n",
    "        new_features.append(feature)\n",
    "\n",
    "    new_seq=Seq(new_sequence) #getting new sequence\n",
    "\n",
    "    genbank.seq=new_seq\n",
    "    genbank.features=new_features\n",
    "    \n",
    "    \n",
    "    start=0\n",
    "    end=0\n",
    "    for feature in genbank.features:\n",
    "        if feature.type ==\"repeat_region\":\n",
    "            if \"TRL\" in feature.qualifiers[\"note\"][0]:\n",
    "                start=int(feature.location.end.position)\n",
    "            elif \"TRS\" in feature.qualifiers[\"note\"][0]:\n",
    "                end=int(feature.location.start.position)\n",
    "                \n",
    "    fastaseq=genbank.seq[start-1:end] #truncating sequence by excluding terminal repeats\n",
    "             \n",
    "                \n",
    "    \n",
    "\n",
    "    outfile=open(sample+\".gb\",\"w\")\n",
    "\n",
    "    SeqIO.write(genbank,outfile,\"genbank\") #writing new genbankfile\n",
    "    \n",
    "    outfile.close()\n",
    "    return str(fastaseq) #return truncated genome sequence\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Script 2:\n",
    "Analysing vcf data for further usage:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import vcf\n",
    "import os\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "from Bio import SeqIO,Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "\n",
    "nucl=(\"A\",\"T\",\"C\",\"G\") #DNA bases\n",
    "\n",
    "out_SNPs=\"sample;REF;POS;ALT;changetype;gene;aachange;AF;DP;dS_or_dN\\n\" #SNP table header\n",
    "out_gene_freq=\"sample\"\n",
    "\n",
    "record=SeqIO.read(\"HSV-1_F-BAC.gb\",\"genbank\") #annotated reference genome\n",
    "\n",
    "gene_list={}\n",
    "gene_syn={}\n",
    "gene_nons={}\n",
    "\n",
    "for feature in record.features:\n",
    "    if feature.type==\"CDS\":\n",
    "        if \"R\" not in feature.qualifiers[\"gene\"][0]:\n",
    "            gene_list[feature.qualifiers[\"gene\"][0]]=0\n",
    "            gene_syn[feature.qualifiers[\"gene\"][0]]=0\n",
    "            gene_nons[feature.qualifiers[\"gene\"][0]]=0\n",
    "            out_gene_freq=out_gene_freq+\";\"+feature.qualifiers[\"gene\"][0]\n",
    "            \n",
    "        \n",
    "out_gene_freq=out_gene_freq+\"\\n\" #gene frequency table header\n",
    "out_dNdS=out_gene_freq\n",
    "out_EGFP=\"sample;EGFP frequency\\n\"\n",
    "\n",
    "csv_file=csv.reader(open(\"samples_min50DP.csv\"), delimiter=\";\") #table with samples and where to find them\n",
    "samples_folder=[]\n",
    "\n",
    "for sample in csv_file: #loading samplelist\n",
    "    samples_folder.append(sample)\n",
    "\n",
    "SNPs=[] #future SNP list\n",
    "gene_freq=[] #future gene frequency list\n",
    "EGFP_freq=[] #future EGFP frequency list\n",
    "namelist=[] #future sample list, replacing Rep1 samplenames\n",
    "dNdS=[]\n",
    "for sample in samples_folder:\n",
    "    \n",
    "    vcf_file=vcf.Reader(filename=sample[1]+\"/vcf_files/\"+sample[0]+\"_lofreq.vcf.gz\",compressed=\"True\") #loading vcf specified by sample and folder\n",
    "    name=\"\"\n",
    "    if \"Rep\" not in sample[0]: #rebranding Rep1 samples because they are not stated as such\n",
    "        name=sample[0]+\"Rep1\"\n",
    "    else:\n",
    "        name=sample[0]\n",
    "        \n",
    "    namelist.append(name)\n",
    "    \n",
    "    for SNP in vcf_file: #loading SNPs\n",
    "        for ALT in SNP.ALT: #for lofreq calls not necessary but just in case there are multiple variants per nucleotide site\n",
    "            \n",
    "            if SNP.INFO[\"AF\"]>=0.05: #only using SNPs with allele frequencies higher than 5%\n",
    "                change=\"NONC\"\n",
    "                for feature in record.features:                \n",
    "                    if feature.type == \"CDS\": \n",
    "                        if feature.location.start < SNP.POS and feature.location.end >= SNP.POS: #checking for mutations within coding sequences\n",
    "                            trans_old=\"\"\n",
    "                            trans_new=\"\"\n",
    "                            change=\"SYN\"\n",
    "                            for segment in feature.location.parts: #in case of multiple parts\n",
    "                                trans_old=trans_old+record.seq[segment.start:segment.end]\n",
    "                                if segment.start<SNP.POS and segment.end>=SNP.POS:\n",
    "                                    trans_new=trans_new+record.seq[segment.start:SNP.POS-1]+str(ALT)+record.seq[SNP.POS:segment.end]\n",
    "                                else:\n",
    "                                    trans_new=trans_new+record.seq[segment.start:segment.end]\n",
    "                            if feature.location.strand==-1:\n",
    "                                trans_old=trans_old.reverse_complement()\n",
    "                                trans_new=trans_new.reverse_complement()\n",
    "                            trans_old=trans_old.translate()\n",
    "                            trans_new=trans_new.translate()\n",
    "                            gene=feature.qualifiers[\"gene\"][0] #note: check how nametags are qulified in your genbank file!\n",
    "                            if trans_old!=trans_new:\n",
    "                                i=1\n",
    "                                c=0\n",
    "                                for old,new in zip(trans_old,trans_new):\n",
    "                                    if old!=new:\n",
    "                                        change=old+str(i)+new\n",
    "\n",
    "                                        c=c+1\n",
    "                                    i=i+1\n",
    "                                if c>1:\n",
    "                                    print(\"fail\",feature.qualifiers['gene'][0]) #something went wrong if there are more than one changes per SNP per CDS e.g. INDELs\n",
    "                            syn=0\n",
    "                            non=0\n",
    "                            for base in nucl: #checking for SYN and NONS sites for this SNP\n",
    "                                mRNA=\"\"\n",
    "                                trans_old=\"\"\n",
    "                                if base!=SNP.REF:\n",
    "                                    for segment in feature.location.parts:\n",
    "                                        trans_old=trans_old+record.seq[segment.start:segment.end]\n",
    "                                        if segment.start<SNP.POS and segment.end>=SNP.POS:\n",
    "                                            mRNA=mRNA+record.seq[segment.start:SNP.POS-1]+base+record.seq[SNP.POS:segment.end]\n",
    "                                        else:\n",
    "                                            mRNA=mRNA+record.seq[segment.start:segment.end]\n",
    "                                    if feature.location.strand==-1:\n",
    "                                        mRNA=mRNA.reverse_complement()\n",
    "                                        trans_old=trans_old.reverse_complement()\n",
    "                                    prot=mRNA.translate()\n",
    "                                    trans_old=trans_old.translate()\n",
    "                                    if prot==trans_old:\n",
    "                                        syn+=1\n",
    "                                    else:\n",
    "                                        non+=1\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "                if change==\"NONC\": #line in case of non coding SNPs\n",
    "                    line=[name,SNP.REF,SNP.POS,ALT,change,\"-\",\"-\",SNP.INFO[\"AF\"],SNP.INFO[\"DP\"],\"-\"]\n",
    "                elif change==\"SYN\": #line in case of synonymous SNPs\n",
    "                    line=[name,SNP.REF,SNP.POS,ALT,change,gene,\"-\",SNP.INFO[\"AF\"],SNP.INFO[\"DP\"],str(SNP.INFO[\"AF\"]/syn)]\n",
    "                else: #line in case of non synonymous SNPs\n",
    "                    line=[name,SNP.REF,SNP.POS,ALT,\"NONS\",gene,change,SNP.INFO[\"AF\"],SNP.INFO[\"DP\"],str(SNP.INFO[\"AF\"]/non)]\n",
    "\n",
    "                SNPs.append(line)\n",
    "                \n",
    "              #calculating gene frequencies from per site depth data  \n",
    "    cov=[]            \n",
    "    for line in (open(sample[1]+\"/coverage/\"+sample[0]+\"_cov_av.txt\")):\n",
    "        cov.append(float(line.split()[2]))\n",
    "    av=0.0\n",
    "    for site in cov:\n",
    "        av+=site\n",
    "    av=av/len(cov)\n",
    "    for feature in record.features:\n",
    "        if feature.type==\"CDS\":\n",
    "            gene_cov=cov[feature.location.start-1:feature.location.end]\n",
    "            gene_av=0.0\n",
    "            for site in gene_cov:\n",
    "                gene_av+=site\n",
    "            gene_av=gene_av/len(gene_cov)\n",
    "            gene_av=gene_av/av\n",
    "            if \"R\" not in feature.qualifiers[\"gene\"][0]:\n",
    "                gene_list[feature.qualifiers[\"gene\"][0]]=gene_av\n",
    "\n",
    "    line=[name]\n",
    "    for gene in gene_list:\n",
    "        line.append(gene_list[gene])\n",
    "    gene_freq.append(line)\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    for a, e in zip(csv.reader(open(sample[1]+\"/coverage/depth_av.csv\"),delimiter=\";\"),csv.reader(open(sample[1]+\"/coverage/depth_EGFP.csv\"),delimiter=\";\")):\n",
    "        if (a[0] and e[0]) == sample[0]:\n",
    "            EGFP_freq.append((name,float(e[1])/float(a[1])))\n",
    "            \n",
    "\n",
    "for sample in namelist:\n",
    "    line=[sample]\n",
    "    for SNP in SNPs:\n",
    "        if (SNP[0]==sample) and (SNP[4]!=\"NONC\") and (SNP[2]!=64393) and (SNP[2]!=64394) and (SNP[2]!=64395) and (\"R\" not in SNP[5]):\n",
    "            if SNP[4]==\"SYN\":\n",
    "                gene_syn[SNP[5]]+=float(SNP[9])\n",
    "            elif SNP[4]==\"NONS\":\n",
    "                gene_nons[SNP[5]]+=float(SNP[9])\n",
    "    for gene in gene_list:\n",
    "        if gene_syn[gene]==0.0:\n",
    "            gene_nons[gene]=gene_nons[gene]/(0.01/3)\n",
    "        else:\n",
    "            gene_nons[gene]=gene_nons[gene]/gene_syn[gene]\n",
    "            \n",
    "        line.append(gene_nons[gene])\n",
    "        gene_syn[gene]=0\n",
    "        gene_nons[gene]=0\n",
    "    dNdS.append(line)\n",
    "    \n",
    "for d,f,e in zip(dNdS,gene_freq,EGFP_freq):\n",
    "    line_d=\"\"\n",
    "    line_f=\"\"\n",
    "    line_e=\"\"\n",
    "    \n",
    "    for el in d:\n",
    "        line_d=line_d+str(el)+\";\"\n",
    "    line_d=line_d[:-1]+\"\\n\"\n",
    "    \n",
    "    for el in f:\n",
    "        line_f=line_f+str(el)+\";\"\n",
    "    line_f=line_f[:-1]+\"\\n\"\n",
    "    \n",
    "    for el in e:\n",
    "        line_e=line_e+str(el)+\";\"\n",
    "    line_e=line_e[:-1]+\"\\n\"\n",
    "    \n",
    "    out_dNdS=out_dNdS+line_d\n",
    "    out_gene_freq=out_gene_freq+line_f\n",
    "    out_EGFP=out_EGFP+line_e\n",
    "    \n",
    "file_dNdS=open(\"dNdS.csv\",\"w\")\n",
    "file_gene_freq=open(\"gene_freq.csv\",\"w\")\n",
    "file_EGFP=open(\"EGFP.csv\",\"w\")\n",
    "\n",
    "file_dNdS.write(out_dNdS[:-1])\n",
    "file_gene_freq.write(out_gene_freq[:-1])\n",
    "file_EGFP.write(out_EGFP[:-1])\n",
    "\n",
    "\n",
    "file_dNdS.close()\n",
    "file_gene_freq.close()\n",
    "file_EGFP.close()\n",
    "\n",
    "\n",
    "for SNP in SNPs:\n",
    "    line=\"\"\n",
    "    for el in SNP:\n",
    "        line=line+str(el)+\";\"\n",
    "    line=line[:-1]+\"\\n\"\n",
    "    \n",
    "    out_SNPs=out_SNPs+line\n",
    "    \n",
    "file_SNPs=open(\"SNPs.csv\",\"w\")\n",
    "file_SNPs.write(out_SNPs[:-1])\n",
    "file_SNPs.close()\n",
    "            \n",
    "    \n",
    "            \n",
    "\n",
    "            \n",
    "            \n",
    "    \n",
    "            \n",
    "                       "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Script 3: \n",
    "Draw dN/dS data onto the genome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "from Bio import SeqIO,Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from reportlab.lib import colors\n",
    "from reportlab.lib.units import cm\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "\n",
    "file=open(\"dNdS.csv\") #open dNdS data generated from the cell above\n",
    "csv=csv.reader(file, delimiter=\";\")\n",
    "dNdS=[]\n",
    "genelist={}\n",
    "for line in csv:\n",
    "    if line[0]==\"sample\":\n",
    "        for gene in line[1:]:\n",
    "            genelist[gene]=0\n",
    "    else:\n",
    "        dNdS.append(line)\n",
    "    \n",
    "samples_woRep=[\"wtACVPXXX\",\"wtFOSPXXX\",\"wtGCVPXXX\",\"wtH2OPXXX\",\"YSACVPXXX\",\"YSFOSPXXX\",\"YSGCVPXXX\",\"YSH2OPXXX\"] #samples for which to calculate dN/dS ratios\n",
    "    \n",
    "record=SeqIO.read(\"HSV-1_F-BAC.gb\",\"genbank\") #open reference file\n",
    "\n",
    "\n",
    "gd_diagram = GenomeDiagram.Diagram(\"HSV-1 F strain\") #genome diagram\n",
    "gd_track_for_features = gd_diagram.new_track(1, name=\"genome\", start=0, end=len(record.seq), scale=0, scale_ticks=0,greytrack=False, greytrack_labels=3) #first track, just including the genome\n",
    "gd_feature_set = gd_track_for_features.new_set()\n",
    "\n",
    "start_old=0\n",
    "start_new=0\n",
    "strand_old=\"fw\"\n",
    "strand_new=\"fw\"\n",
    "for feature in record.features:\n",
    "    if feature.type==\"CDS\":\n",
    "        label=False\n",
    "        if feature.location.strand==1:\n",
    "            strand_new=\"fw\"\n",
    "            start_new=int(feature.location.start.position)\n",
    "        else:\n",
    "            strand_new=\"rev\"\n",
    "            start_new=int(feature.location.end.position)\n",
    "        if (start_new-start_old<1500) and (strand_new==strand_old):\n",
    "            label=False\n",
    "        else:\n",
    "            label=True\n",
    "            \n",
    "        start_old=start_new\n",
    "        strand_old=strand_new\n",
    "        \n",
    "        if \"U\" in feature.qualifiers[\"gene\"][0]:\n",
    "            color=\"grey\"\n",
    "        elif \"R\" in feature.qualifiers[\"gene\"][0]:\n",
    "            color=\"silver\"\n",
    "        gd_feature_set.add_feature(feature, color=color, label=label, label_position=\"middle\", label_angle=90, sigil=\"ARROW\", arrowhead_length=0.2, arrowshaft_height=1.0)\n",
    "\n",
    "for selection in samples_woRep: #checking per selection\n",
    "    gd_track_for_features = gd_diagram.new_track(3, name=selection, start=0, end=len(record.seq), scale=1, scale_ticks=0,greytrack=True, greytrack_labels=2)\n",
    "    gd_feature_set = gd_track_for_features.new_set()\n",
    "    for sample in dNdS:\n",
    "        if sample[0][:-4]==selection: #sample features selection\n",
    "            for ratio,gene in zip(sample[1:],genelist):\n",
    "                if float(ratio)>2.0: #only including genes with dN/dS ratios above 2.0\n",
    "                    genelist[gene]+=1\n",
    "    for gene in genelist:\n",
    "        for feature in record.features:\n",
    "            if feature.type ==\"CDS\":\n",
    "                if feature.qualifiers[\"gene\"][0]==gene:\n",
    "                    if genelist[gene]>=2: #plotting if at least 2 out of 3 replicates show dN/dS ratios above 2.0\n",
    "                        if \"wt\" in selection:\n",
    "                            color=\"lightcoral\"\n",
    "                        elif \"YS\" in selection:\n",
    "                            color=\"cornflowerblue\"\n",
    "                        label=False\n",
    "                        if genelist[gene]==3: #labeling gene if 3 out of 3 replicates with dN/dS above 2.0\n",
    "                            label=True\n",
    "                        if feature.location.strand==1:\n",
    "                            angle=0\n",
    "                            position=\"start\"\n",
    "                        else:\n",
    "                            angle=180\n",
    "                            position=\"end\"\n",
    "                        gd_feature_set.add_feature(feature, color=color, label=label, label_angle=angle, sigil=\"ARROW\", arrowhead_length=0.2, arrowshaft_height=1.0,label_position=position)\n",
    "    for gene in genelist:\n",
    "        genelist[gene]=0\n",
    "\n",
    "        \n",
    "gd_track_for_features = gd_diagram.new_track(11, name=\"legend\", start=0, end=len(record.seq), scale=0, scale_ticks=0,greytrack=False, greytrack_labels=3)\n",
    "gd_feature_set = gd_track_for_features.new_set() #Legendtrack\n",
    "\n",
    "pink=(SeqFeature(FeatureLocation(30000, 34000),1))\n",
    "lightblue=(SeqFeature(FeatureLocation(110000, 114000),1))\n",
    "\n",
    "gd_feature_set.add_feature(pink, color=\"lightcoral\", label=True, name=\"dN/dS >2 for at least 2 wt populations\", label_angle=0, sigil=\"ARROW\", arrowhead_length=0.2, arrowshaft_height=1.0, label_position=\"end\")\n",
    "gd_feature_set.add_feature(lightblue, color=\"cornflowerblue\", label=True, name=\"dN/dS >2 for at least 2 YS populations\", label_angle=0, sigil=\"ARROW\", arrowhead_length=0.2, arrowshaft_height=1.0, label_position=\"end\")\n",
    "\n",
    "        \n",
    "gd_diagram.draw(\n",
    "    format=\"linear\",\n",
    "    orientation=\"landscape\",\n",
    "    pagesize=(25*cm,10*cm),\n",
    "    fragments=1,\n",
    "    start=0,\n",
    "    end=len(record.seq)\n",
    ")\n",
    "\n",
    "gd_diagram.write(\"dNdS2ormore.pdf\", \"PDF\")\n",
    "                \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Script 4:\n",
    "Principal component analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import csv\n",
    "\n",
    "df=pd.read_csv(\"dNdS.csv\",header=0,sep=\";\") #PCA of dN/dS data, different data can be analyzed though\n",
    "\n",
    "\n",
    "\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "features = []\n",
    "for col in df.columns:\n",
    "    features.append(col)\n",
    "features=features[1:]\n",
    "# Separating out the features\n",
    "x = df.loc[:, features].values\n",
    "# Separating out the target\n",
    "y = df.loc[:,[\"sample\"]].values\n",
    "# Standardizing the features\n",
    "x = StandardScaler().fit_transform(x)\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "pca = PCA(n_components=2) #pca with 2 components\n",
    "principalComponents = pca.fit_transform(x)\n",
    "principalDf = pd.DataFrame(data = principalComponents, columns = ['principal component 1', 'principal component 2'])\n",
    "\n",
    "\n",
    "targets=[]\n",
    "for sample in csv.reader(open(\"dNdS.csv\",\"r\"),delimiter=\";\"):\n",
    "    if sample[0]!=\"sample\":\n",
    "        targets.append(sample[0]) #exclude header\n",
    "\n",
    "finalDf = pd.concat([principalDf, df[['sample']]], axis = 1)\n",
    "fig = plt.figure(figsize = (4,4))\n",
    "ax = fig.add_subplot(1,1,1) \n",
    "ax.set_xlabel('Principal Component 1', fontsize = 12)\n",
    "ax.set_ylabel('Principal Component 2', fontsize = 12)\n",
    "#ax.set_title('2 component PCA', fontsize = 20)\n",
    "finalDf\n",
    "\n",
    "transper={\"P0\":0.05,\"PV\":0.1,\"PX\":0.25,\"PXV\":0.45,\"PXX\":0.65,\"PXXV\":0.85,\"PXXX\":1.0} #giving different passages different transperacy values\n",
    "for target in targets:\n",
    "    if \"GCV\" in target:\n",
    "        marker=\"^\"\n",
    "    elif \"FOS\" in target:\n",
    "        marker=\"s\"\n",
    "    elif \"ACV\" in target:\n",
    "        marker=\"o\"\n",
    "    elif \"H2O\" in target:\n",
    "        marker=\"h\"\n",
    "        \n",
    "    if \"PXXXRep2\" in target:\n",
    "        label=target[:2]+\" \"+target[2:target.find(\"P\")]\n",
    "    else:\n",
    "        label=\"_nolegend_\"\n",
    "        \n",
    "    if \"wt\" in target:\n",
    "        color=\"lightcoral\"\n",
    "    else:\n",
    "        color=\"cornflowerblue\"\n",
    "        \n",
    "    indicesToKeep = finalDf['sample'] == target\n",
    "\n",
    "    scatter=ax.scatter(finalDf.loc[indicesToKeep, 'principal component 1']\n",
    "               , finalDf.loc[indicesToKeep, 'principal component 2']\n",
    "               , s = 30\n",
    "               ,c=color\n",
    "               ,alpha=transper[target[target.find(\"P\"):target.find(\"Rep\")]]\n",
    "               ,label=label\n",
    "               ,marker=marker)\n",
    "\n",
    "#plt.xlim(-2.5, 2)\n",
    "#plt.ylim(-2.5,3)\n",
    "plt.tight_layout()\n",
    "legend1=ax.legend( loc='best', frameon=False, fontsize=8)\n",
    "ax.add_artist(legend1)\n",
    "\n",
    "plt.savefig(\"PCA_all.pdf\")\n",
    "plt.show()\n",
    "\n",
    "finalDf.to_csv(\"PCA.csv\")\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Script 5:\n",
    "Calculate allele frequencies over passaging time:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import csv\n",
    "\n",
    "\n",
    "SNPs=[]\n",
    "samples=[]\n",
    "for SNP in csv.reader(open(\"SNPs.csv\",\"r\"),delimiter=\";\"):\n",
    "    if \"sample\" not in SNP[0]:\n",
    "        SNPs.append(SNP)\n",
    "        if \"P0\" not in SNP[0]:\n",
    "            sample=SNP[0][:SNP[0].find(\"P\")]+SNP[0][SNP[0].find(\"R\"):]\n",
    "            if sample not in samples:\n",
    "                samples.append(sample)\n",
    "AF=[]\n",
    "\n",
    "for SNP in SNPs:\n",
    "    if \"P0\" not in SNP[0]:\n",
    "        line=[SNP[0][:SNP[0].find(\"P\")]+SNP[0][SNP[0].find(\"R\"):],SNP[1],SNP[2],SNP[3],SNP[4],SNP[5],SNP[6],\"-\",\"-\",\"-\",\"-\",\"-\",\"-\",\"-\"]\n",
    "        if line not in AF:\n",
    "            AF.append(line)\n",
    "    elif (\"wt\" in SNP[0]) or (\"YS\" in SNP[0]):\n",
    "        for sample in samples:\n",
    "            if SNP[0][:2]==sample[:2]:\n",
    "                line=[sample,SNP[1],SNP[2],SNP[3],SNP[4],SNP[5],SNP[6],\"-\",\"-\",\"-\",\"-\",\"-\",\"-\",\"-\"]\n",
    "                if line not in AF:\n",
    "                    AF.append(line)\n",
    "    \n",
    "passage={\"P0\":7,\n",
    "         \"PV\":8,\n",
    "         \"PX\":9,\n",
    "         \"PXV\":10,\n",
    "         \"PXX\":11,\n",
    "         \"PXXV\":12,\n",
    "         \"PXXX\":13}        \n",
    "for SNP in AF:\n",
    "    for snp in SNPs:\n",
    "        pas=\"\"\n",
    "        if (SNP[0]==snp[0][:snp[0].find(\"P\")]+snp[0][snp[0].find(\"R\"):]):\n",
    "            pas=snp[0][snp[0].find(\"P\"):snp[0].find(\"R\")]\n",
    "            if (SNP[1:6]==snp[1:6]):\n",
    "                SNP[passage[pas]]=snp[7]\n",
    "        elif (\"P0\" in snp[0]) and (SNP[0][:2]==snp[0][:2]):\n",
    "            pas=\"P0\"\n",
    "            if (SNP[1:6]==snp[1:6]):\n",
    "                SNP[passage[pas]]=snp[7]\n",
    "out_AF=\"sample;REF;POS;ALT;changetype;gene;aachange;AF P0;AF PV;AF PX;AF PXV;AF PXX;AF PXXV;AF PXXX\\n\"            \n",
    "for SNP in AF:\n",
    "    line=\"\"\n",
    "    for el in SNP:\n",
    "        line=line+str(el)+\";\"\n",
    "    line=line[:-1]+\"\\n\"\n",
    "    out_AF=out_AF+line\n",
    "outfile=open(\"AF_overtime.csv\",\"w\")\n",
    "outfile.write(out_AF[:-2])\n",
    "outfile.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Script 6: Plot SNPs over time:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import csv\n",
    "import numpy as np\n",
    "\n",
    "fig, axs = plt.subplots(ncols=2, nrows=4, sharex=True, sharey=True, figsize=(12, 8))\n",
    "\n",
    "mut={\"wt\":0, \"YS\":1}\n",
    "sel={\"ACV\":0, \"FOS\":1, \"GCV\":2, \"H2O\":3}\n",
    "SNPs=[]\n",
    "samples=[]\n",
    "passage_number=[5,10,15,20,25,30]\n",
    "for SNP in csv.reader(open(\"AF_overtime.csv\",\"r\"),delimiter=\";\"):\n",
    "    if \"sample\" not in SNP[0]:\n",
    "        SNPs.append(SNP)\n",
    "        if SNP[0][:SNP[0].find(\"R\")] not in samples:\n",
    "            samples.append(SNP[0][:SNP[0].find(\"R\")])\n",
    "\n",
    "\n",
    "legend=[]\n",
    "for sample in samples:\n",
    "    pos=[-0.14]\n",
    "    c=0\n",
    "    gs=[]\n",
    "    Mut=sample[:2]\n",
    "    Sel=sample[2:]\n",
    "    \n",
    "    for SNP in SNPs:\n",
    "        if (SNP[5] ==\"UL30\") and (SNP[0][:SNP[0].find(\"R\")]==sample):\n",
    "            if SNP[7]==\"-\":\n",
    "                y=[0]\n",
    "            else:\n",
    "                y=[float(SNP[7])]\n",
    "            x=[0]\n",
    "            for freq, pas in zip(SNP[8:],passage_number):\n",
    "                x.append(float(pas))\n",
    "                if freq!=\"-\":\n",
    "                    y.append(float(freq))\n",
    "                else:\n",
    "                    y.append(0.0)\n",
    "                \n",
    "            if \"FOS\" in SNP[0]:\n",
    "                marker=\"s\"\n",
    "            elif \"ACV\" in SNP[0]:\n",
    "                marker=\"o\"\n",
    "            elif \"H2O\" in SNP[0]:\n",
    "                marker=\"h\"\n",
    "            elif \"GCV\" in SNP[0]:\n",
    "                marker=\"^\"\n",
    "\n",
    "            if \"wt\" in SNP[0]:\n",
    "                color=\"lightcoral\"\n",
    "                alpha=0.7\n",
    "            elif \"YS\" in SNP[0]:\n",
    "                color=\"cornflowerblue\"\n",
    "                alpha=0.7\n",
    "            if sample not in legend:\n",
    "                legend.append(sample)\n",
    "                name=sample\n",
    "            else:\n",
    "                name=\"_nolegend_\"\n",
    "                \n",
    "                \n",
    "            axs[sel[Sel],mut[Mut]].plot(x,y,linewidth=2,marker=marker,color=color,alpha=alpha, label=name)\n",
    "            if (SNP[4]==\"NONS\"):\n",
    "                if (SNP[5]==\"UL30\") and (SNP[6]==\"Y557N\") :\n",
    "                    name=\"Y557S\"\n",
    "                else:\n",
    "                    name=SNP[6]\n",
    "            else:\n",
    "                name=SNP[1]+SNP[2]+SNP[3]\n",
    "                \n",
    "            \n",
    "                \n",
    "            i=0    \n",
    "            for position in pos:\n",
    "                \n",
    "                if abs(position-y[-1])<0.08:\n",
    "                    i=i+1\n",
    "            if x[-1]==30:\n",
    "                pos.append(y[-1])\n",
    "                if i==0:\n",
    "                    if y[-1]>0.05:\n",
    "                        axs[sel[Sel],mut[Mut]].text(x[-1]+2,y[-1],name,fontsize=7)\n",
    "            if x[-1]==30:\n",
    "                c+=1\n",
    "    axs[sel[Sel],mut[Mut]].plot([0,30],[0.05,0.05],linewidth=3,color=\"grey\",alpha=1.0, linestyle=\"--\")\n",
    "                   \n",
    "        \n",
    "\n",
    "    \n",
    "axs[3,0].set_xlabel('Passage',fontsize=12)\n",
    "axs[3,1].set_xlabel('Passage',fontsize=12)\n",
    "axs[0,0].set_ylabel('Allele Frequency',fontsize=12)\n",
    "axs[1,0].set_ylabel('Allele Frequency',fontsize=12)\n",
    "axs[2,0].set_ylabel('Allele Frequency',fontsize=12)\n",
    "axs[3,0].set_ylabel('Allele Frequency',fontsize=12)\n",
    "\n",
    "fig.legend(frameon=True,ncol=4,bbox_to_anchor=(0.45, 0.45))\n",
    "fig.suptitle(\"UL30 Mutations\", fontsize=20)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"UL30_ot.pdf\")\n",
    "\n",
    "plt.show()\n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Script 7:\n",
    "Ploting SNPs on the genome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "from Bio import SeqIO,Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from reportlab.lib import colors\n",
    "from reportlab.lib.units import cm\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "\n",
    "file=open(\"SNPs.csv\")\n",
    "csv=csv.reader(file, delimiter=\";\")\n",
    "SNPs=[]\n",
    "samples_woRep=[]\n",
    "genelist={}\n",
    "for line in csv:\n",
    "    if line[0]!=\"sample\":\n",
    "        SNPs.append(line)\n",
    "    \n",
    "samples_woRep=[\"wtACV\",\"wtFOS\",\"wtGCV\",\"wtH2O\",\"YSACV\",\"YSFOS\",\"YSGCV\",\"YSH2O\"]\n",
    "    \n",
    "record=SeqIO.read(\"HSV-1_F-BAC_Pol.gb\",\"genbank\")\n",
    "\n",
    "\n",
    "gd_diagram = GenomeDiagram.Diagram(\"HSV-1 F strain\")\n",
    "gd_track_for_features = gd_diagram.new_track(1, name=\"genome\", start=0, end=len(record.seq), scale=0, scale_ticks=0,greytrack=False, greytrack_labels=3)\n",
    "gd_feature_set = gd_track_for_features.new_set()\n",
    "\n",
    "start_old=0\n",
    "start_new=0\n",
    "strand_old=\"fw\"\n",
    "strand_new=\"fw\"\n",
    "for feature in record.features:\n",
    "    if feature.type==\"CDS\":\n",
    "        label=False\n",
    "        if feature.location.strand==1:\n",
    "            strand_new=\"fw\"\n",
    "            start_new=int(feature.location.start.position)\n",
    "        else:\n",
    "            strand_new=\"rev\"\n",
    "            start_new=int(feature.location.end.position)\n",
    "        if (start_new-start_old<1500) and (strand_new==strand_old):\n",
    "            label=False\n",
    "        else:\n",
    "            label=True\n",
    "            \n",
    "        start_old=start_new\n",
    "        strand_old=strand_new\n",
    "        \n",
    "        shaft=0.2\n",
    "        if \"U\" in feature.qualifiers[\"gene\"][0]:\n",
    "            color=\"grey\"\n",
    "            shaft=0.2\n",
    "        elif \"R\" in feature.qualifiers[\"gene\"][0]:\n",
    "            color=\"silver\"\n",
    "        elif \"Exo\" in feature.qualifiers[\"gene\"][0]:\n",
    "            color=\"blue\"\n",
    "            shaft=0.85\n",
    "            label=True\n",
    "        else:\n",
    "            color=\"orange\"\n",
    "            shaft=0.85\n",
    "            label=True\n",
    "        if feature.location.strand==1:\n",
    "            label_angle=0\n",
    "        else:\n",
    "            label_angle=180\n",
    "            \n",
    "        if \"Exo\" in feature.qualifiers[\"gene\"][0]:\n",
    "            label_angle=90\n",
    "        \n",
    "        if \"site\" in feature.qualifiers[\"gene\"][0]:\n",
    "            label_angle=135\n",
    "            \n",
    "     \n",
    "        label=True\n",
    "        gd_feature_set.add_feature(feature, color=color, label=label, label_position=\"start\", label_angle=label_angle, sigil=\"ARROW\", arrowhead_length=0.2, arrowshaft_height=shaft)\n",
    "    \n",
    "for selection in samples_woRep:\n",
    "    gd_track_for_features = gd_diagram.new_track(3, name=selection, start=0, end=len(record.seq), scale=1, scale_ticks=0,greytrack=True, greytrack_labels=4)\n",
    "    gd_feature_set = gd_track_for_features.new_set()\n",
    "    features=[]\n",
    "    for SNP in SNPs:\n",
    "        if (SNP[0][:SNP[0].find(\"PXXX\")]==selection) and (SNP[5] == \"UL23\"):\n",
    "            \n",
    "            if SNP[4]==\"NONS\":\n",
    "                name=str(SNP[6])\n",
    "                if SNP[6]==\"Y557N\":\n",
    "                    name=\"Y557S\"\n",
    "                else:\n",
    "                    name=str(SNP[6])\n",
    "            elif SNP[4]==\"SYN\":\n",
    "                name=str(SNP[1])+str(SNP[2])+str(SNP[3])\n",
    "                \n",
    "            \n",
    "                \n",
    "            mut=SeqFeature(FeatureLocation(int(SNP[2]),int(SNP[2])+1))\n",
    "            \n",
    "            if \"wt\" in SNP[0]:\n",
    "                color=\"lightcoral\"\n",
    "            elif \"YS\" in SNP[0]:\n",
    "                color=\"cornflowerblue\"\n",
    "            i=0\n",
    "            for feature in features:\n",
    "                if abs(feature.location.start-int(SNP[2]))<80:\n",
    "                    i=i+1\n",
    "            if i==0:\n",
    "                label=True\n",
    "            else:\n",
    "                label=False\n",
    "                \n",
    "        \n",
    "            \n",
    "                   \n",
    "            start_old=start_new\n",
    "            gd_feature_set.add_feature(mut,name=name,label=label, label_angel=90, sigil=\"BOX\",color=color)\n",
    "            features.append(mut)\n",
    "\n",
    "                           \n",
    "gd_diagram.draw(\n",
    "    format=\"linear\",\n",
    "    orientation=\"landscape\",\n",
    "    pagesize=(25*cm,8*cm),\n",
    "    fragments=1,\n",
    "    start=46589,\n",
    "    end=47719\n",
    ")\n",
    "\n",
    "gd_diagram.write(\"UL23_genome.pdf\", \"PDF\")\n",
    "                \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Script 8:\n",
    "Ploting coverage data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import csv\n",
    "import numpy as np\n",
    "\n",
    "fig, axs = plt.subplots(ncols=2, nrows=4, sharex=True, sharey=True, figsize=(12, 8))\n",
    "mut={\"wt\":0, \"YS\":1}\n",
    "sel={\"ACV\":0, \"FOS\":1, \"GCV\":2, \"H2O\":3}\n",
    "csv_file=csv.reader(open(\"samples_min50DP.csv\"), delimiter=\";\") #table with samples and where to find them\n",
    "samples_folder=[]\n",
    "legend=[]\n",
    "for sample in csv_file: #loading samplelist\n",
    "    samples_folder.append(sample)\n",
    "    \n",
    "for sample in samples_folder:\n",
    "    if \"PXXX\" in sample[0]:\n",
    "        Mut=sample[0][:2]\n",
    "        Sel=sample[0][2:sample[0].find(\"P\")]\n",
    "        cov_file=open(sample[1]+\"/coverage/\"+sample[0]+\"_cov_av.txt\")\n",
    "        av=0.0\n",
    "        cov=[]\n",
    "        for site in cov_file:\n",
    "            av=av+int(site.split()[2])\n",
    "            cov.append(site.split())\n",
    "        av=av/len(cov)\n",
    "        x=[]\n",
    "        y=[]\n",
    "        for site in cov:\n",
    "            if (int(site[1])>=46589) and (int(site[1])<=47719):\n",
    "                x.append(int(site[1]))\n",
    "                y.append(float(site[2])/av)\n",
    "        if \"FOS\" in sample[0]:\n",
    "            marker=\"s\"\n",
    "        elif \"ACV\" in sample[0]:\n",
    "            marker=\"o\"\n",
    "        elif \"H2O\" in sample[0]:\n",
    "            marker=\"h\"\n",
    "        elif \"GCV\" in sample[0]:\n",
    "            marker=\"^\"\n",
    "\n",
    "        if \"wt\" in sample[0]:\n",
    "            color=\"lightcoral\"\n",
    "            alpha=0.7\n",
    "        elif \"YS\" in sample[0]:\n",
    "            color=\"cornflowerblue\"\n",
    "            alpha=0.7\n",
    "\n",
    "        if \"Rep\" not in sample[0]:\n",
    "            sample[0]=sample[0]+\"Rep1\"\n",
    "        \n",
    "        if sample[0][:sample[0].find(\"P\")] not in legend:\n",
    "            legend.append(sample[0][:sample[0].find(\"P\")])\n",
    "            name=sample[0][:sample[0].find(\"P\")]\n",
    "        else:\n",
    "            name=\"_nolegend_\"\n",
    "                \n",
    "                \n",
    "        axs[sel[Sel],mut[Mut]].plot(x,y,linewidth=2,marker=marker,color=color,alpha=alpha, label=name)\n",
    "        \n",
    "axs[3,0].set_xlabel('Position',fontsize=12)\n",
    "axs[3,1].set_xlabel('Position',fontsize=12)\n",
    "axs[0,0].set_ylabel('Coverage',fontsize=12)\n",
    "axs[1,0].set_ylabel('Coverage',fontsize=12)\n",
    "axs[2,0].set_ylabel('Coverage',fontsize=12)\n",
    "axs[3,0].set_ylabel('Coverage',fontsize=12)\n",
    "fig.legend(frameon=True,ncol=4,bbox_to_anchor=(0.45, 0.82))\n",
    "fig.suptitle(\"UL23 Coverage\", fontsize=20)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"UL23Cov.pdf\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Script 9:\n",
    "Calculating and ploting SNP counts:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import csv\n",
    "\n",
    "\n",
    "SNPs=[]\n",
    "samples=[]\n",
    "samples2=[]\n",
    "passage_number=[0,5,10,15,20,25,30]\n",
    "\n",
    "fig = plt.figure(figsize = (6,3.7))\n",
    "\n",
    "for SNP in csv.reader(open(\"SNPs.csv\",\"r\"),delimiter=\";\"):\n",
    "    if \"sample\" not in SNP[0]:\n",
    "        SNPs.append(SNP)\n",
    "        if SNP[0] not in samples:\n",
    "            samples.append(SNP[0])\n",
    "        if \"P0\" not in SNP[0]:\n",
    "            sample=SNP[0][:SNP[0].find(\"P\")]+SNP[0][SNP[0].find(\"R\"):]\n",
    "            if sample not in samples2:\n",
    "                samples2.append(sample)\n",
    "AF=[]\n",
    "for SNP in SNPs:\n",
    "        if \"P0\" not in SNP[0]:\n",
    "            line=[SNP[0][:SNP[0].find(\"P\")]+SNP[0][SNP[0].find(\"R\"):],\"-\",\"-\",\"-\",\"-\",\"-\",\"-\",\"-\"]\n",
    "            if line not in AF:\n",
    "                AF.append(line)\n",
    "        elif (\"wt\" in SNP[0]) or (\"YS\" in SNP[0]):\n",
    "            for sample in samples2:\n",
    "                if SNP[0][:2]==sample[:2]:\n",
    "                    line=[sample,\"-\",\"-\",\"-\",\"-\",\"-\",\"-\",\"-\"]\n",
    "                    if line not in AF:\n",
    "                        AF.append(line)\n",
    "            \n",
    "passage={\"P0\":1,\n",
    "         \"PV\":2,\n",
    "         \"PX\":3,\n",
    "         \"PXV\":4,\n",
    "         \"PXX\":5,\n",
    "         \"PXXV\":6,\n",
    "         \"PXXX\":7}              \n",
    "\n",
    "for sample in samples:\n",
    "    SNPcount=0\n",
    "    for SNP in SNPs:\n",
    "        if SNP[0]==sample:\n",
    "            SNPcount+=1  \n",
    "    for SNP in AF:\n",
    "        pas=\"\"\n",
    "        if (SNP[0]==sample[:sample.find(\"P\")]+sample[sample.find(\"R\"):]):\n",
    "            pas=sample[sample.find(\"P\"):sample.find(\"R\")]\n",
    "            SNP[passage[pas]]=SNPcount\n",
    "        elif (\"P0\" in sample) and (SNP[0][:2]==sample[:2]):\n",
    "            pas=\"P0\"\n",
    "            SNP[passage[pas]]=SNPcount\n",
    "out_AF=\"sample;AF P0;AF PV;AF PX;AF PXV;AF PXX;AF PXXV;AF PXXX\\n\"            \n",
    "for SNP in AF:\n",
    "    line=\"\"\n",
    "    for el in SNP:\n",
    "        line=line+str(el)+\";\"\n",
    "    line=line[:-1]+\"\\n\"\n",
    "    out_AF=out_AF+line\n",
    "outfile=open(\"SNPcount_overtime.csv\",\"w\")\n",
    "outfile.write(out_AF[:-1])\n",
    "outfile.close()\n",
    "\n",
    "for sample in AF:\n",
    "    if sample[0]!=\"sample\":\n",
    "        if \"Rep2\" in sample[0]:\n",
    "            label=sample[0][:sample[0].find(\"R\")]\n",
    "        else:\n",
    "            label=\"_nolegend_\"\n",
    "            \n",
    "        x=[]\n",
    "        y=[]\n",
    "        for freq, pas in zip(sample[1:],passage_number):\n",
    "            if freq!=\"-\":\n",
    "                x.append(float(pas))\n",
    "                y.append(float(freq))\n",
    "        if \"FOS\" in sample[0]:\n",
    "            marker=\"s\"\n",
    "        elif \"ACV\" in sample[0]:\n",
    "            marker=\"o\"\n",
    "        elif \"H2O\" in sample[0]:\n",
    "            marker=\"h\"\n",
    "        elif \"GCV\" in sample[0]:\n",
    "            marker=\"^\"\n",
    "\n",
    "        if \"wt\" in sample[0]:\n",
    "            color=\"lightcoral\"\n",
    "            alpha=0.7\n",
    "        elif \"YS\" in sample[0]:\n",
    "            color=\"cornflowerblue\"\n",
    "            alpha=0.7\n",
    "\n",
    "        plt.plot(x,y,linewidth=2,marker=marker,color=color,label=label,alpha=alpha)\n",
    "\n",
    "\n",
    "plt.xlabel(\"Passage\",fontsize=12)\n",
    "plt.ylabel(\"SNP count\",fontsize=12)\n",
    "plt.legend(frameon=False, ncol=2)\n",
    "plt.xlim(0,30.5)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"SNP_count.pdf\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Script 10:\n",
    "Generating clone genbank files + multiple FASTA files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "import os\n",
    "from vcftogenbankDef import vcftogenbank #if in different script\n",
    "\n",
    "\n",
    "depth=csv.reader(open(\"coverage/depth_av.csv\",\"r\"),delimiter=\";\") #coverage/depth_av.csv file includes all samplenames + coverage data\n",
    "samplelist=[]\n",
    "coveragedic={}\n",
    "fastadic={}\n",
    "for el in depth:\n",
    "    if \"name\" not in el[0]:\n",
    "        samplelist.append(el[0])\n",
    "        coveragedic[el[0]]=float(el[1])\n",
    "\n",
    "for sample in samplelist:\n",
    "    if coveragedic[sample]>30.0:\n",
    "        fastadic[sample]=vcftogenbank(\"HSV-1_F-BAC.gb\",sample) #see script 1 in this notebook, check if samplenames are comparable\n",
    "wtGCV=\"\"\n",
    "YSGCV=\"\"\n",
    "wtH2O=\"\"\n",
    "YSH2O=\"\"\n",
    "for sample in fastadic:\n",
    "    if \"wtGCV\" in sample:\n",
    "        wtGCV=wtGCV+\">\"+sample+\"\\n\"+fastadic[sample]+\"\\n\"\n",
    "    elif \"YSGCV\" in sample:\n",
    "        YSGCV=YSGCV+\">\"+sample+\"\\n\"+fastadic[sample]+\"\\n\" \n",
    "    elif \"wtH2O\" in sample:\n",
    "        wtH2O=wtH2O+\">\"+sample+\"\\n\"+fastadic[sample]+\"\\n\"\n",
    "    elif \"YSH2O\" in sample:\n",
    "        YSH2O=YSH2O+\">\"+sample+\"\\n\"+fastadic[sample]+\"\\n\" \n",
    "\n",
    "wtGCVfile=open(\"wtGCV.fasta\",\"w\")\n",
    "YSGCVfile=open(\"YSGCV.fasta\",\"w\")\n",
    "wtH2Ofile=open(\"wtH2O.fasta\",\"w\")\n",
    "YSH2Ofile=open(\"YSH2O.fasta\",\"w\")\n",
    "\n",
    "wtGCVfile.write(wtGCV)\n",
    "YSGCVfile.write(YSGCV)\n",
    "wtH2Ofile.write(wtH2O)\n",
    "YSH2Ofile.write(YSH2O)\n",
    "\n",
    "wtGCVfile.close()\n",
    "YSGCVfile.close()\n",
    "wtH2Ofile.close()\n",
    "YSH2Ofile.close()\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Script 11:\n",
    "Heterozygosity Calculation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import vcf\n",
    "import os\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "from Bio import SeqIO,Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "record=SeqIO.read(\"HSV-1_F-BAC.gb\",\"genbank\") #annotated reference genome\n",
    "\n",
    "gene_list={}\n",
    "out=\"population;\"\n",
    "for feature in record.features:\n",
    "        if feature.type==\"CDS\":\n",
    "            if \"R\" not in feature.qualifiers[\"gene\"][0]:\n",
    "                gene_list[feature.qualifiers[\"gene\"][0]]={}\n",
    "                out=out+feature.qualifiers[\"gene\"][0]+\";\"\n",
    "                \n",
    "                \n",
    "out=out[:-1]+\"\\n\"\n",
    "\n",
    "\n",
    "        \n",
    "            \n",
    "population=(\"wtGCV\",\"wtH2O\",\"YSGCV\",\"YSH2O\")\n",
    "clone=(\"1\",\"2\",\"3\",\"4\",\"5\",\"6\",\"7\",\"8\",\"9\",\"10\",\"11\",\"12\",\"13\",\"14\",\"15\",\"16\",\"17\",\"18\",\"19\",\"20\")\n",
    "\n",
    "for pop in population:\n",
    "    out=out+pop+\";\"\n",
    "    for feature in record.features:\n",
    "        if feature.type==\"CDS\":\n",
    "            if \"R\" not in feature.qualifiers[\"gene\"][0]:\n",
    "                gene_list[feature.qualifiers[\"gene\"][0]]={}\n",
    "                \n",
    "    \n",
    "    for c in clone:\n",
    "            if os.path.exists(\"Run300622/fasta_gb/\"+pop+\"c\"+c+\".gb\"):\n",
    "                gb=SeqIO.read(\"Run300622/fasta_gb/\"+pop+\"c\"+c+\".gb\",\"genbank\") #open respective genbank file\n",
    "                \n",
    "                \n",
    "                for feature in gb.features:\n",
    "                    if feature.type==\"CDS\":\n",
    "                        if \"R\" not in feature.qualifiers[\"gene\"][0]:\n",
    "                            gene=feature.qualifiers[\"gene\"][0]\n",
    "                            allele=gb.seq[feature.location.start-1:feature.location.end]\n",
    "                            if allele in gene_list[gene]:\n",
    "                                gene_list[gene][allele]+=1\n",
    "                            else:\n",
    "                                gene_list[gene][allele]=1\n",
    "    hetero={}\n",
    "    for gene in gene_list:\n",
    "        pop_size=0\n",
    "        hetero[gene]=1\n",
    "        for allele in gene_list[gene]:\n",
    "            pop_size+=gene_list[gene][allele]\n",
    "        for allele in gene_list[gene]:\n",
    "            gene_list[gene][allele]=float(gene_list[gene][allele]/pop_size)\n",
    "            hetero[gene]-=np.square(gene_list[gene][allele])\n",
    "        out=out+str(hetero[gene])+\";\"\n",
    "    out=out[:-1]+\"\\n\"\n",
    "    \n",
    "out=out[:-1]   \n",
    "#print(out)\n",
    "outfile=open(\"heterozygosity.csv\",\"w\")\n",
    "outfile.write(out)\n",
    "\n",
    "outfile.close()\n",
    "        \n",
    "            \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
